File: snippets/go/develop/skillsmd/actions.go
package develop

import (
	"fmt"
	"math/rand/v2"
	"time"

	restate "github.com/restatedev/sdk-go"
)

type Actions struct{}

func (Actions) StateOperations(ctx restate.ObjectContext, greeting string) error {
	// <start_state>
	myString := "my-default"
	if s, err := restate.Get[*string](ctx, "my-string-key"); err != nil {
		return err
	} else if s != nil {
		myString = *s
	}

	count, err := restate.Get[int](ctx, "count")
	if err != nil {
		return err
	}

	// Set state
	restate.Set(ctx, "my-key", "my-new-value")
	restate.Set(ctx, "count", count+1)

	// Clear state
	restate.Clear(ctx, "my-key")
	restate.ClearAll(ctx)
	stateKeys, err := restate.Keys(ctx)
	// <end_state>

	if err != nil {
		return err
	}
	_ = stateKeys

	_ = myString
	return nil
}

func (Actions) ServiceCommunication(ctx restate.Context, greeting string) error {
	request := "Hi"
	objectKey := "object-key"
	workflowId := "wf-id"

	// <start_service_calls>
	// Call a Service
	svcResponse, err := restate.Service[string](ctx, "MyService", "MyHandler").
		Request(request)
	if err != nil {
		return err
	}

	// Call a Virtual Object
	objResponse, err := restate.Object[string](ctx, "MyObject", objectKey, "MyHandler").
		Request(request)
	if err != nil {
		return err
	}

	// Call a Workflow
	wfResponse, err := restate.Workflow[string](ctx, "MyWorkflow", workflowId, "Run").
		Request(request)
	if err != nil {
		return err
	}
	// <end_service_calls>

	_ = svcResponse
	_ = objResponse
	_ = wfResponse
	return nil
}

func (Actions) OneWayMessages(ctx restate.Context, greeting string) error {
	request := "Hi"
	objectKey := "object-key"
	workflowId := "wf-id"

	// <start_sending_messages>
	// Send to service
	restate.ServiceSend(ctx, "MyService", "MyHandler").Send(request)

	// Send to virtual object
	restate.ObjectSend(ctx, "MyObject", objectKey, "MyHandler").Send(request)

	// Send to workflow
	restate.WorkflowSend(ctx, "MyWorkflow", workflowId, "Run").Send(request)
	// <end_sending_messages>

	return nil
}

func (Actions) DelayedMessages(ctx restate.Context, greeting string) error {
	request := "Hi"

	// <start_delayed_messages>
	restate.ServiceSend(ctx, "MyService", "MyHandler").Send(request, restate.WithDelay(5*time.Hour))
	// <end_delayed_messages>

	return nil
}

func (Actions) DurableSteps(ctx restate.Context, greeting string) error {
	// <start_durable_steps>
	result, err := restate.Run(ctx, func(ctx restate.RunContext) (string, error) {
		return callExternalAPI(), nil
	}, restate.WithName("Call to API"))
	if err != nil {
		return err
	}
	// <end_durable_steps>

	// <start_async_run>
	call1 := restate.RunAsync(ctx, func(ctx restate.RunContext) (string, error) {
		return callExternalAPI(), nil
	})
	user, err := call1.Result()
	// <end_async_run>

	// <start_deterministic_helpers>
	// Deterministic UUID
	uuid := restate.UUID(ctx)

	// Deterministic random numbers
	randomInt := restate.Rand(ctx).Uint64()
	randomFloat := restate.Rand(ctx).Float64()

	// Use as a math/rand/v2 source
	mathRandV2 := rand.New(restate.RandSource(ctx))

	// time
	now, err := restate.Run(ctx, func(ctx restate.RunContext) (time.Time, error) {
		return time.Now(), nil
	})
	// <end_deterministic_helpers>
	_ = now
	_ = uuid
	_ = randomInt
	_ = randomFloat
	_ = mathRandV2
	_ = user
	_ = result
	return nil
}

func (Actions) DurableTimers(ctx restate.Context, greeting string) error {
	// <start_durable_timers>
	err := restate.Sleep(ctx, 30*time.Second)
	if err != nil {
		return err
	}
	// <end_durable_timers>

	// <start_async_sleep>
	sleepFuture := restate.After(ctx, 30*time.Second)
	// ... do other work ...
	if err := sleepFuture.Done(); err != nil {
		return err
	}
	// <end_async_sleep>

	return nil
}

func (Actions) Awakeables(ctx restate.Context, greeting string) error {
	name := "Pete"

	// <start_awakeables>
	awakeable := restate.Awakeable[string](ctx)
	awakeableId := awakeable.Id()

	// Send ID to external system
	if _, err := restate.Run(ctx, func(ctx restate.RunContext) (string, error) {
		return requestHumanReview(name, awakeableId), nil
	}); err != nil {
		return err
	}

	// Wait for result
	review, err := awakeable.Result()
	// <end_awakeables>
	if err != nil {
		return err
	}

	// <start_awakeables_resolution>
	restate.ResolveAwakeable(ctx, awakeableId, "Looks good!")
	restate.RejectAwakeable(ctx, awakeableId, fmt.Errorf("Cannot be reviewed"))
	// <end_awakeables_resolution>

	_ = review
	return nil
}

func (Actions) WorkflowPromises(ctx restate.WorkflowContext, greeting string) error {
	// <start_workflow_promises>
	// Wait for promise
	promise := restate.Promise[string](ctx, "review")
	review, err := promise.Result()
	if err != nil {
		return err
	}

	// Resolve promise from another handler
	err = restate.Promise[string](ctx, "review").Resolve(review)
	if err != nil {
		return err
	}
	// <end_workflow_promises>

	_ = review
	return nil
}

func (Actions) InvocationManagement(ctx restate.Context, greeting string) error {
	// <start_cancel>
	invocationId := restate.
		ServiceSend(ctx, "MyService", "MyHandler").
		// Optional: send attaching idempotency key
		Send("Hi", restate.WithIdempotencyKey("my-idempotency-key")).
		GetInvocationId()

	// Later re-attach to the request
	response, err := restate.AttachInvocation[string](ctx, invocationId).Response()
	if err != nil {
		return err
	}

	// I don't need this invocation anymore, let me just cancel it
	restate.CancelInvocation(ctx, invocationId)
	// <end_cancel>

	// <start_wait_first>
	sleepFuture := restate.After(ctx, 30*time.Second)
	callFuture := restate.Service[string](ctx, "MyService", "MyHandler").RequestFuture("hi")

	fut, err := restate.WaitFirst(ctx, sleepFuture, callFuture)
	// <end_wait_first>
	_ = fut
	_ = response

	return nil
}

func (Actions) Combinators(ctx restate.Context, greeting string) (string, error) {
	// <start_wait_all>
	callFuture1 := restate.Service[string](ctx, "MyService", "MyHandler").RequestFuture("hi")
	callFuture2 := restate.Service[string](ctx, "MyService", "MyHandler").RequestFuture("hi again")

	for fut, err := range restate.Wait(ctx, callFuture1, callFuture2) {
		if err != nil {
			return "", err
		}
		resp, _ := fut.(restate.ResponseFuture[string]).Response()
		return resp, nil
		// process response
	}
	// <end_wait_all>

	// <start_idempotency>
	invocationId := restate.ServiceSend(ctx, "MyService", "MyHandler").
		Send("Hi", restate.WithIdempotencyKey("my-key")).
		GetInvocationId()
	// <end_idempotency>

	// <start_attach>
	response, err := restate.AttachInvocation[string](ctx, invocationId).Response()
	// <end_attach>
	_ = response

	// <start_cancel>
	restate.CancelInvocation(ctx, invocationId)
	// <end_cancel>
	return "", err
}

// Helper functions
func callExternalAPI() string {
	return "external result"
}

func requestHumanReview(name, awakeableId string) string {
	// External review logic
	return "review-id"
}

File: snippets/java/src/main/java/develop/skillsmd/Actions.java
package develop.skillsmd;

import dev.restate.common.InvocationOptions;
import dev.restate.common.Request;
import dev.restate.common.Target;
import dev.restate.sdk.*;
import dev.restate.sdk.common.DurablePromiseKey;
import dev.restate.sdk.common.StateKey;
import dev.restate.serde.TypeRef;
import dev.restate.serde.TypeTag;
import develop.MyObject;
import develop.MyService;
import develop.MyWorkflow;
import java.time.Duration;
import java.util.Collection;
import java.util.List;
import java.util.UUID;

public class Actions {

  // <start_generic_state>
  // import dev.restate.serde.TypeRef;
  private static final StateKey<List<String>> ITEMS =
      StateKey.of("items", new TypeRef<List<String>>() {});

  // <end_generic_state>

  public void stateOperations() {
    // <start_state>
    var state = Restate.state();
    StateKey<String> STRING_STATE_KEY = StateKey.of("my-key", String.class);
    String stringState = state.get(STRING_STATE_KEY).orElse("my-default");
    state.set(STRING_STATE_KEY, "my-new-value");
    state.clear(STRING_STATE_KEY);
    state.clearAll();
    Collection<String> keys = state.getAllKeys();
    // <end_state>
  }

  public void serviceCommunication() {
    String request = "Hi";
    String objectKey = "object-key";
    String workflowId = "wf-id";

    // <start_service_calls>
    String svcResponse =
        Restate.serviceHandle(MyService.class).call(MyService::myHandler, request).await();
    String objResponse =
        Restate.virtualObjectHandle(MyObject.class, objectKey)
            .call(MyObject::myHandler, request)
            .await();
    String wfResponse =
        Restate.workflowHandle(MyWorkflow.class, workflowId).call(MyWorkflow::run, request).await();
    // <end_service_calls>
  }

  public void genericServiceCalls() {
    String request = "Hi";

    // <start_generic_calls>
    // Define target
    Target target = Target.service("MyService", "myHandler");
    Target objectTarget = Target.virtualObject("MyObject", "object-key", "myHandler");
    Target workflowTarget = Target.workflow("MyWorkflow", "wf-id", "run");

    // Do the call
    String response =
        Restate.call(
                Request.of(target, TypeTag.of(String.class), TypeTag.of(String.class), request))
            .await();

    // Or send the message
    Restate.send(Request.of(target, TypeTag.of(String.class), TypeTag.of(String.class), request));
    // <end_generic_calls>
  }

  public void oneWayMessages() {
    String request = "Hi";
    String objectKey = "Hi";
    String workflowId = "Hi";

    // <start_sending_messages>
    Restate.serviceHandle(MyService.class).send(MyService::myHandler, request);
    Restate.virtualObjectHandle(MyObject.class, objectKey).send(MyObject::myHandler, request);
    Restate.workflowHandle(MyWorkflow.class, workflowId).send(MyWorkflow::run, request);
    // <end_sending_messages>
  }

  public void delayedMessages() {
    String request = "Hi";

    // <start_delayed_messages>
    Restate.serviceHandle(MyService.class).send(MyService::myHandler, request, Duration.ofDays(5));
    // <end_delayed_messages>
  }

  public void durableSteps() {
    // <start_durable_steps>
    // Wrap non-deterministic code in Restate.run
    String result = Restate.run("call external API", String.class, () -> callExternalAPI());

    // Wrap with name for better tracing
    String namedResult = Restate.run("my-side-effect", String.class, () -> callExternalAPI());
    // <end_durable_steps>

    // <start_deterministic_helpers>
    float value = Restate.random().nextFloat();
    UUID uuid = Restate.random().nextUUID();
    // <end_deterministic_helpers>

  }

  public void durableTimers() {
    // <start_durable_timers>
    Restate.sleep(Duration.ofHours(30));
    // <end_durable_timers>
  }

  public void awakeables() {
    String name = "Pete";

    // <start_awakeables>
    // Create awakeable
    Awakeable<String> awakeable = Restate.awakeable(String.class);
    String awakeableId = awakeable.id();

    // Send ID to external system
    Restate.run("request-human-review", () -> requestHumanReview(name, awakeableId));

    // Wait for result
    String review = awakeable.await();
    // <end_awakeables>

    // <start_awakeables_resolution>
    Restate.awakeableHandle(awakeableId).resolve(String.class, "Looks good!");
    Restate.awakeableHandle(awakeableId).reject("Cannot be reviewed");
    // <end_awakeables_resolution>
  }

  public void workflowPromises() {
    // <start_workflow_promises>
    DurablePromiseKey<String> REVIEW_PROMISE = DurablePromiseKey.of("review", String.class);
    // Wait for promise
    String review = Restate.promise(REVIEW_PROMISE).future().await();

    // Resolve promise from another handler
    Restate.promiseHandle(REVIEW_PROMISE).resolve(review);
    // <end_workflow_promises>
  }

  public void concurrency() {
    // <start_combine_all>
    // Wait for all to complete
    DurableFuture<String> call1 =
        Restate.serviceHandle(MyService.class).call(MyService::myHandler, "request1");
    DurableFuture<String> call2 =
        Restate.serviceHandle(MyService.class).call(MyService::myHandler, "request2");

    DurableFuture.all(call1, call2).await();
    // <end_combine_all>

    // <start_combine_any>
    String res = Select.<String>select().or(call1).or(call2).await();
    // <end_combine_any>
  }

  public void invocationManagement() {
    var request = "";

    // <start_idempotency>
    var handle =
        Restate.serviceHandle(MyService.class)
            .send(MyService::myHandler, request, InvocationOptions.idempotencyKey("abc123"));
    // <end_idempotency>

    // <start_attach>
    var response = handle.attach().await();
    // <end_attach>

    // <start_cancel>
    handle.cancel();
    // <end_cancel>
  }

  // Helper methods
  private String callExternalAPI() {
    return "external result";
  }

  private void requestHumanReview(String name, String awakeableId) {
    // External review logic
  }
}
